### PR TITLE
Add a public_file_prefix option to cashier.conf

### LINKED ISSUE #49: The lack of files causes problems.
OK, this was a longer post, but with some testing discovered there's a resolution that won't require `cashier` to write the private key/cert. It only needs to write the public key and the public certificate to files.

So the problem:

Say you have two ssh CAs. One signs keys for `*.prod.acme-corp.com` and the other signs them for `*.test.acme-corp.com`.  With the traditional world of unsigned keys, you'd do something like this in your `.ssh/config`:

```
Host *.prod.acme-corp.com
  IdentityFile ~/.ssh/id_rsa_prod

Host *.test.acme-corp.com
  IdentityFile ~/.ssh/id_rsa_test
```

Normally it's a single key per device. But with signed keys, each set of servers covered by an ssh CA gets a key/cert. If we get to a world where services like Gitlab and Github are supporting signed keys the multiple key thing becomes an issue.

With the `IdentitiesOnly` ssh option, you can force ssh to pick a specific key with `IdentityFile` (and `CertificateFile` which we'd need for signed keys). And these options can use the public key.

The `ssh-add -d` command has the same issue - it also wants a filename and it too can take a public key (or public cert).

A `cashier` option to save the public part of the signed key/cert files allows `IdentitiesOnly` to work.

### PR DESCRIPTION
**Pull request by lyda**
Allow the client to save the public key and public cert to files
that start with public_file_prefix and end with .pub and -cert.pub
respectively.

This is the naming scheme the ssh IdentityFile config option supported
for certs starting in version 5.4p1. Starting in version 7.2p1, an
additional option, CertificateFile, was added, but the IdentityFile-only
method with those names still works.

Used in conjunction with a user's ~/.ssh/config file setting
IdentitiesOnly and IdentityFile, this change will allow for multiple
ssh CAs for different services.

Note that this will resolve #49 .

### CODE REVIEW

File: client/config.go [hunk 1]
```diff
@@ -3,6 +3,8 @@ package client
 import (
 	"github.com/spf13/pflag"
 	"github.com/spf13/viper"
+	"os/user"
```
Comment (nsheridan) [inline client/config.go]: https://github.com/mitchellh/go-homedir?

Oh now that I've looked ahead I see that you're expanding `~user/blah` as well as `~/blah` and I think go-homedir only expands the `~/` case

File: client/config.go [hunk 2]
```diff
@@ -12,16 +14,33 @@ type Config struct {
...
 	viper.SetDefault("validateTLSCertificate", true)
 }
 
+// expandTilde expands ~ and ~user for a given path.
+func expandTilde(path string) string {
+	re := regexp.MustCompile("^~([^/]*)(/.*)")
+	if m := re.FindStringSubmatch(path); len(m) > 0 {
+		u, _ := user.Current()
+		if m[1] != "" {
+			u, _ = user.Lookup(m[1])
```
Comment (nsheridan) [inline client/config.go]: You're chucking the error here. What happens if the user doesn't exist? If this is called with `expandTilde("~omg/blah")` and `omg` doesn't exist I think what's going to happen is that `"~omg/blah"` will be returned and later code will write some files in the current directory with the names `~omg/blah.pub` and `~omg/blah-cert.pub`.

File: client/client.go [hunk 3]
```diff
@@ -16,6 +18,24 @@ import (
 	"golang.org/x/crypto/ssh/agent"
 )
 
+// InstallPublicFiles installs the public part of the cert and key.
+func InstallPublicFiles(prefix string, cert *ssh.Certificate, pub ssh.PublicKey) error {
```
Comment (nsheridan) [inline client/client.go]: I think this is totally up for lawyering but maybe name this `SavePublicFiles()`? I realise `InstallCert` isn't a great name either 😄

File: README.md [hunk 4]
```diff
@@ -242,26 +242,49 @@ Vault support is currently a work-in-progress.
...
 - A new ssh private key. Generate one in the usual way using `ssh-keygen -f ssh_ca` - this is your CA signing key. At this time Cashier supports RSA, ECDSA and Ed25519 keys. *Important* This key should be kept safe - *ANY* ssh key signed with this key will be able to access your machines.
 - OAuth (Google or GitHub) credentials. You may also need to set the callback URL when creating these.
 
 ## Using cashier
-Once the server is up and running you'll need to configure your client.  
-The client is configured using either a [HCL](https://github.com/hashicorp/hcl) configuration file - [example](example-client.conf) - or command-line flags.  
+Once the server is up and running you'll need to configure your client.
+The client is configured using either a [HCL](https://github.com/hashicorp/hcl) configuration file - [example](example-client.conf) - or command-line flags.
+
+- `ca`          CA server (default "http://localhost:10000").
```
Comment (nsheridan) [inline README.md]: Nice touch, adding the flags. Nitpick - maybe add `--`? e.g. `--ca`, `--config`...

File: README.md [hunk 5]
```diff
@@ -242,26 +242,49 @@ Vault support is currently a work-in-progress.
...
+- `config`      Path to config file (default "/Users/kevin/.cashier.conf").
```
Comment (nsheridan) [inline README.md]: The hell it is :D

File: README.md [hunk 6]
```diff
@@ -242,26 +242,49 @@ Vault support is currently a work-in-progress.
...
+- `key_size`    Key size. Ignored for ed25519 keys (default 2048).
+- `key_type`    Type of private key to generate - rsa, ecdsa or ed25519 (default "rsa").
+- `public_file_prefix` Prefix for filename for public key and cert (optional, no default). The public key is put in a file with `.pub` appended to it; the public cert file in a file with `-cert.pub` appended to it.
+- `validity`    Key validity (default 24h0m0s).
```
Comment (nsheridan) [inline README.md]: I know these came from the usage string, but maybe shorten this to just `24h` for easier reading.

File: README.md [hunk 7]
```diff
@@ -153,8 +153,8 @@ server {
 }
 ```
 
-Prior to using MySQL, MongoDB or SQLite you need to create the database and tables using [one of the provided files](db).
```
Comment (lyda) [inline README.md]: Ah. I just kinda mindlessly so `:%s/ . *$//` when I notice trailing whitespace in a file due to experience with some other, let's say less careful, developers. I'll put those back, sorry.

Was unaware that was a thing in markdown. Honestly if it was me I'd just put an explicit `<br />` in there or a `\n\n` just because I can't see trailing spaces. I'll see about changing my vim config to make those visible if that's a thing people do.

File: README.md (hunk 4 above)
Comment (lyda) [inline README.md]: Done.

File: README.md (hunk 5 above)
Comment (lyda) [inline README.md]: lol! Yes, this was added with `!!./cashier --help` and I clearly didn't review it! Sigh. Thanks for that catch.

File: README.md (hunk 6 above)
Comment (lyda) [inline README.md]: Done.

File: README.md [hunk 8]
```diff
@@ -242,26 +242,49 @@ Vault support is currently a work-in-progress.
...
 Running the `cashier` cli tool will open a browser window at the configured CA address.
-The CA will redirect to the auth provider for authorisation, and redirect back to the CA where the access token will printed.  
-Copy the access token. In the terminal where you ran the `cashier` cli paste the token at the prompt.  
-The client will then generate a new ssh key-pair and send the public part to the server (along with the access token).  
+The CA will redirect to the auth provider for authorisation, and redirect back to the CA where the access token will printed.
+Copy the access token. In the terminal where you ran the `cashier` cli paste the token at the prompt.
+The client will then generate a new ssh key-pair and send the public part to the server (along with the access token).
 Once signed the client will install the key and signed certificate in your ssh agent. When the certificate expires it will be removed automatically from the agent.
 
+If you set `public_file_prefix` then the public key and public cert
```
Comment (lyda) [inline README.md]: Done, though GFM doesn't put `<br />`'s after single newlines (unlike traditional Markdown which does) so the rendered results look the same. I tend to wrap under 80 so it's easy to read on a terminal, but I'm elderly. Renders well either way in the browser which is generally where it's read.

File: client/client.go [hunk 9]
```diff
@@ -16,6 +18,24 @@ import (
...
+	if prefix == "" {
+		return nil
+	}
+	pubKey := prefix + ".pub"
+	pubCert := prefix + "-cert.pub"
```
Comment (lyda) [inline client/client.go]: Ah, yes. That was a lot longer when I was passing in the full config.

The same is true for `pubTxt` and `certPubTxt` but the WriteFile lines become kind of dense at that point. The compiler is doing the same thing either way so I'd prefer keeping that to make it a bit easier to read.

File: client/client.go [hunk 10]
```diff
@@ -16,6 +18,24 @@ import (
...
+	pubTxt := ssh.MarshalAuthorizedKey(pub)
+	certPubTxt := []byte(cert.Type() + " " + base64.StdEncoding.EncodeToString(cert.Marshal()))
```
Comment (lyda) [inline client/client.go]: Yep. Looked at `strings.Join` for that reason, but it's three strings and it's not in a loop so this way seemed like an easier read.

File: client/config.go (hunk 1 above)
Comment (lyda) [inline client/config.go]: Yes, but is `~user` really needed? I went hunting for a tilde expanding lib but instead saw lots of posts from people saying not to do it. Which is fine, but ssh's `IdentityFile` supports it and the obvious thing for people to do is make `IdentityFile` match `public_file_prefix`.

I don't know if `IdentityFile` supports `~user` syntax, but the one person in the world who does that can maintain his own cashier version.

Lib added, code nuked.

File: client/config.go (hunk 2 above)
Comment (lyda) [inline client/config.go]: Code gone, so moot. But yes, should have coughed up an error here.

File: client/config.go (hunk 1 above)
Comment (lyda) [inline client/config.go]: Note `cmd/cashier/main.go` is the only remaining non-test user of os/user. I'll change that in a later PR though.

File: client/client.go (hunk 3 above)
Comment (lyda) [inline client/client.go]: I like `Save` more, yes. Was just trying to make them match. As for `InstallCert` I can't think of a better term. `UploadCertToAgent` maybe but then you'd have to port this all to Java and isn't there enough horribleness in the world?

**Comment by lyda** [comment]
OK, now fixed.

Would it be OK if I committed a script for the travis checks that one could also use as a pre-commit hook?

File: README.md (hunk 7 above)
Comment (nsheridan) [inline README.md]: Yeah, I'll tidy it up at some point. Trailing whitespace to denote meaning. Wat.

File: client/client.go (hunk 10 above)
Comment (nsheridan) [inline client/client.go]: 👍

File: client/config.go (hunk 1 above)
Comment (nsheridan) [inline client/config.go]: Yeah I have no idea why anyone who would  want to write files to the homedir of a different user would get annoyed enough at the prospect of providing a full path to fork 😄 
I've been wrong before though

**Comment by nsheridan** [comment]
I'm not sure what the point of a code review tool which requires a person with write access to perform merges is, even if the code review is positive.
